compiler-hacking.md

```diff
@@ -0,0 +1,502 @@
+concrete class have less since they just execute the behaviour defined by the
+abstract class.
+
+For general naming conventions:
+- _resolve_ methods convert strings into useful metadata (_e.g._ find the [`FunctionDefinition`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/FunctionDefinition.java) given the name in the script)
+- _render_ methods generate bytecode
+- classes that end in _Node_ are generate type of syntax in the language; their subclasses represent concrete bits of syntax the user can write
+- classes that end in _Builder_ are utility classes that make generating bytecode simpler
+
+`ExpressionNode` has many similar clones. For instance, there is an expression
+node for `For` and then ones for each of the sources, clauses, subsamples, and
+collectors in a `For` expression, containing other expressions.  While these
+are structurally different, they are built very similarly to `ExpressionNode`.
+
+### Type Information
+Computing captures is done as part of bytecode production. The type system for
+Shesmu is meant to be very simple and static. There is no inheritance and no
+automatic conversion. It's also entirely bottom up. For instance, in Java, type
+information flows both ways:
+
+     List<Foo> x = new ArrayList<>();
+
+Type information about the type argument `Foo` is flowing in the opposite
+direction as the type information about `ArrayList<Foo>` being assigned to
+`List<Foo>`. 
+
+In Shesmu, everything flows from the leaves of the tree towards the root. For instance:
+
+     path_file(dir + 'x.txt')
+
+If `dir` cannot be resolved or was defined by an expression that has a type
+error, this has no valid type. Even though we know that `path_file` expects a
+`path` type and the `+` operator can combine two paths. Shesmu will simply give
+up. This is meant to keep the compiler simple.
+
+### Bytecode Generation
+When generating bytecode, Shesmu uses
+[`GeneratorAdapter`](https://asm.ow2.io/javadoc/org/objectweb/asm/commons/GeneratorAdapter.html)
+from the ASM package to write bytecode. It wraps this in a
+[`Renderer`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/Renderer.java)
+class. `Renderer` provides a number of utility functions, such as inserting
+`Imyhat` objects into the bytecode. It's real purpose is to track extra
+information. It knows where the current stream value is stored, which will be
+different through many layers of capturing. It also tracks how to access
+signatures.
+
+### Stream Purity
+Shesmu has two kinds of stream purity: olive-level (type) and `For`-level
+(order). They are conceptually similar, but perform different functions.
+
+For all operations on Java streams, the order or type of the items can be
+changed by some operations.
+
+In the case of olives, `Group`, `Join`, `LeftJoin` change the type of the
+stream value while `Where`, `Monitor`, `Reject`, and `Pick` do not. A `Define`
+olive is written expecting the original type of the input stream. Therefore, no
```

Comment:
How about "A `Define` olive has more stringent requirements; the original type of the input stream (as defined in `Input x;`) is the only valid input type for `Define` olives. Therefore, no [...]"